versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/CommitLogicImpl.java

```diff
@@ -302,14 +300,38 @@ public boolean storeCommit(@Nonnull CommitObj commit, @Nonnull List<Obj> additio
       commit = indexTooBigStoreUpdate(commit);
 
       try {
-        return persist.storeObj(commit, true);
+        return mitigateHashCollision(persist.storeObj(commit, true), commit);
       } catch (ObjTooLargeException ex) {
         // Hit the "Hard database object size limit"
         throw new RuntimeException(ex);
       }
     }
   }
 
+  /**
+   * Called from the above {@link #storeCommit(CommitObj, List)}, handles the case when it could not
+   * persist the {@link CommitObj} (duplicate object-id). Checks whether the persisted object is
+   * actually the object that's expected to be persisted - and yields "OK" in that case. This
+   * mitigates the risk of false-positive hash-collision errors in case the backend database runs
+   * into timeout situations with an undefined outcome.
+   */
+  private CommitObj mitigateHashCollision(boolean storeResult, CommitObj commit) {
+    if (storeResult) {
+      return commit;
+    }
+
+    // Check whether the existing object is the same commit (w/o considering the internal "created"
+    // timestamp of the commit-obj).
+    try {
+      CommitObj existing = persist.fetchTypedObj(commit.id(), COMMIT, CommitObj.class);
+      CommitObj commitWithNewCreatedTimestamp =
+          CommitObj.commitBuilder().from(commit).created(existing.created()).build();
```

Comment:
Would it make sense to mark `create()` as `@Value.Auxiliary`? It should hopefully also make it clear why we do not hash it. WDYT?

Reply:
Why? That's not a marker annotation.